packages/dd-trace/src/encode/agentless-ci-visibility.js

```diff
@@ -0,0 +1,245 @@
+      buffer[offset + 1] = value >> 8
+      buffer[offset + 2] = value
+      return
+    }
+
+    if (value >= -0x80000000) { // int 32
+      bytes.reserve(5)
+      bytes.length += 5
+      buffer[offset] = 0xd2
+      buffer[offset + 1] = value >> 24
+      buffer[offset + 2] = value >> 16
+      buffer[offset + 3] = value >> 8
+      buffer[offset + 4] = value
+      return
+    }
+
+    // int 64
+    const hi = Math.floor(value / Math.pow(2, 32))
+    const lo = value >>> 0
+
+    bytes.reserve(9)
+    bytes.length += 9
+
+    buffer[offset] = 0xd3
+    buffer[offset + 1] = hi >> 24
+    buffer[offset + 2] = hi >> 16
+    buffer[offset + 3] = hi >> 8
+    buffer[offset + 4] = hi
+    buffer[offset + 5] = lo >> 24
+    buffer[offset + 6] = lo >> 16
+    buffer[offset + 7] = lo >> 8
+    buffer[offset + 8] = lo
+  }
+
+  _encodePositiveLong (bytes, value) {
+    const buffer = bytes.buffer
+    const offset = bytes.length
+
+    if (value < 0x80) { // positive fixnum
+      bytes.reserve(1)
+      bytes.length += 1
+      buffer[offset] = value
+      return
+    }
+
+    if (value < 0x100) { // uint 8
+      bytes.reserve(2)
+      bytes.length += 2
+      buffer[offset] = 0xcc
+      buffer[offset + 1] = value
+      return
+    }
+
+    if (value < 0x10000) { // uint 16
+      bytes.reserve(3)
+      bytes.length += 3
+      buffer[offset] = 0xcd
+      buffer[offset + 1] = value >> 8
+      buffer[offset + 2] = value
+      return
+    }
+
+    if (value < 0x100000000) { // uint 32
+      bytes.reserve(5)
+      bytes.length += 5
+      buffer[offset] = 0xce
+      buffer[offset + 1] = value >> 24
+      buffer[offset + 2] = value >> 16
+      buffer[offset + 3] = value >> 8
+      buffer[offset + 4] = value
+      return
+    }
+
+    const hi = (value / Math.pow(2, 32)) >> 0
+    const lo = value >>> 0
+
+    bytes.reserve(9)
+    bytes.length += 9
+
+    buffer[offset] = 0xcf
+    buffer[offset + 1] = hi >> 24
+    buffer[offset + 2] = hi >> 16
+    buffer[offset + 3] = hi >> 8
+    buffer[offset + 4] = hi
+    buffer[offset + 5] = lo >> 24
+    buffer[offset + 6] = lo >> 16
+    buffer[offset + 7] = lo >> 8
+    buffer[offset + 8] = lo
+  }
+
+  _encode (bytes) {
+    const payload = {
+      version: ENCODING_VERSION,
+      metadata: {
+        'language': 'javascript',
+        'library_version': tracerVersion,
+        'runtime.name': 'node',
+        'runtime.version': process.version
+      },
+      events: this._events.map(formatSpan)
+    }
+    if (this.service) {
+      payload.metadata.service = this.service
+    }
+    if (this.env) {
+      payload.metadata.env = this.env
+    }
+    if (this.runtimeId) {
+      payload.metadata['runtime-id'] = this.runtimeId
+    }
+
+    log.debug(() => {
+      return `Adding encoded trace to buffer: ${JSON.stringify(payload)}`
+    })
+
+    this._encodeMap(bytes, payload)
+  }
+
+  makePayload () {
+    const bytes = this._traceBytes
+
+    this._encode(bytes)
```

Comment:
Why is this called here?

Reply:
I should've explained this in the description, my bad. 

So the agentless writer works slightly differently: when a trace is `append`ed, the agentless writer stores it "as is" in the buffer `this._events`. When `makePayload` is called, the encoding of the whole buffer is done "at once".

Reply:
I tried doing it like the current agent exporter but it seemed more complicated because of the payload format

Reply:
It shouldn't be that more complicated. I think it will scale better in the future if this encoder supports buffering the encoded data for when we use it outside of CI. It should be straightforward to write the start of the payload upfront, and then all events as they come in.

Reply:
I did this in https://github.com/DataDog/dd-trace-js/pull/1880/commits/8174339e930a7b199e81bd55aca1b28a397d271f. It is a bit awkward having to "pre-encode" the start of the payload because you need to update it later on when you know the number of events you're going to flush, but at least now it is aligned with 0.4